### ISSUE TITLE
Overlapping glyphs in Times New Roman substitute on Ubuntu 18.04 LTS

### ISSUE THREAD

**Issue by LegionMammal978** [mozilla/pdf.js#13176]
Attach (recommended) or Link to PDF file here: https://www.cityofpaloalto.org/civicax/filebank/documents/8700

Configuration:
- Web browser and its version: Firefox 87.0
- Operating system and its version: Ubuntu 18.04.5 LTS
- PDF.js version: v2.7.368 to v2.8.322
- Is a browser extension: No

Steps to reproduce the problem:
1. Create a fresh installation of Ubuntu 18.04 LTS without any additional fonts.
2. Download and build any version of pdf.js no older than 45d9ab6.
2. Download the PDF and open it using `web/viewer.html`.

What is the expected behavior? (add screenshot) The document should render with an appropriate sans-serif font, since Ubuntu does not have Times New Roman by default. The correct behavior occurs on Ubuntu 20.04 LTS; the first page of the document is shown in [this screenshot](https://user-images.githubusercontent.com/7880963/113496333-ace42380-94c6-11eb-9559-8698fae2cc28.png).

What went wrong? (add screenshot) On Ubuntu 18.04 LTS, the glyphs overlap (or at least connect), rendering the text difficult to read. This is illustrated in [this screenshot](https://user-images.githubusercontent.com/7880963/113496320-7908fe00-94c6-11eb-898f-4ef09a58ee1a.png). Note how the letters are placed identically in the two images; however, the substitute font in 18.04 has a higher weight and ex-height, causing this issue.

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension):

**Comment by LegionMammal978** [comment]
The issue also appears to extend to detection of serif vs. sans-serif fonts. [minimal-times.pdf](https://github.com/mozilla/pdf.js/files/6253866/minimal-times.pdf) attempts to load the system font `/Times#20New#20Roman`, but with `/Flags 32`, indicating a sans-serif font. [minimal-nonexistent.pdf](https://github.com/mozilla/pdf.js/files/6253871/minimal-nonexistent.pdf) attempts to load `/NonExistentFontName` with the same flags. (The font information on these files was derived from a real-world PDF document.) pdf.js on Ubuntu 20.04, as well as Evince and ImageMagick on both OS versions, renders the first file with an appropriate serif font and the second file with the default sans-serif font. However, pdf.js on Ubuntu 18.04 loads the first file with a sans-serif font with overlapping glyphs.

I've done a bisection to find that the issue was originally introduced by commit 45d9ab6. For reference, [this screenshot](https://user-images.githubusercontent.com/7880963/113496618-78be3200-94c9-11eb-8cc3-59707991e45f.png) shows the first page of the legal document on Ubuntu 18.04 using 00b4f86, its parent commit. It can be seen that although the kerning of the word *Purposes* in the top-right corner was corrected, the remainder of the text encounters this issue. 2600e59 attempted to fix certain overlaps caused by the original commit, but it has no effect on this issue.

**Comment by wbv** [comment]
This is still an issue on my machine and has been for some time.

JS/frontend debugging is a bit over my head so I'm afraid I can't add any developer insight, but I can reproduce the issue using the minimal examples provided by @LegionMammal978 (thanks for those!).

### Configuration
- Web browser and its version: Firefox 109.0 (64-bit)
- Operating system and its version: Gentoo Linux (amd64)
- PDF.js version: 3.2.47 (commit feb6f5951)
- Is a browser extension: No

### Steps to reproduce
- Open [minimal-nonexistent.pdf](https://github.com/mozilla/pdf.js/files/6253871/minimal-nonexistent.pdf) in a new tab
- Open [minimal-times.pdf](https://github.com/mozilla/pdf.js/files/6253866/minimal-times.pdf) in a new tab

### What is the expected behavior?
- minimal-nonexistent.pdf shows a nicely-rendered sans-serif font:

![image](https://user-images.githubusercontent.com/7481569/212743807-e6776e66-dac4-45ee-906b-a1c7dfb8b36a.png)

### What went wrong?
- minimal-times.pdf shows the same sans-serif font with kerning fudged to "match" Times, which comes out very ugly (since the glyphs are intended for a different set of font metrics):

![image](https://user-images.githubusercontent.com/7481569/212743857-0cc30a5c-b5e1-493c-9808-fa314bfb53fd.png)

This particular font-selection/metrics mismatch makes reading some PDFs, particularly those generated by Microsoft Word (with Times New Roman fonts) difficult. For example, this conference paper: [https://digitalcommons.usu.edu/cgi/viewcontent.cgi?filename=0&article=4666&context=smallsat&type=additional](https://digitalcommons.usu.edu/cgi/viewcontent.cgi?filename=0&article=4666&context=smallsat&type=additional)

![image](https://user-images.githubusercontent.com/7481569/212745027-7ed32d4d-2e21-49b4-bdd7-c317160843af.png)

**Comment by marco-c** [comment]
#16363 might have fixed this, can you still reproduce in latest Firefox Nightly?

**Comment by LegionMammal978** [comment]
I can still reproduce with the pdf.js version included in latest Firefox Nightly. However, I cannot reproduce with pdf.js built from latest master (e738e15aa397ef91b75fcf952abd0d368b232f20, v3.7.27). I bisected the fix to 53134c0c0bb2d19244dd8c31aa4e8e6d01a657e4 (v3.7.18), which is from the PR you're referring to. Now, both the original document and `minimal-times.pdf` render with an appropriate serif font on Ubuntu 18.04 LTS, with no overlapping glyphs.

**Comment by marco-c** [comment]
Latest Firefox Nightly should contain this fix. Are you sure you are testing on the latest one? (in the about window, you should see 2023-05-12).

**Comment by LegionMammal978** [comment]
No, I was using 2023-05-11; perhaps the PPA I got it from is a bit slow to update.

**Comment by LegionMammal978** [comment]
Alright, I've downloaded Nightly 2023-05-12 directly and confirmed that the issue is fixed there.
